File: reverse_direction/10.有道翻译.js
const crypto = require('crypto');

// sign 加密
function getSign() {
    let timestamp = (new Date).getTime();
    let l = 'fanyideskweb';
    let d = 'webfanyi';
    let e = 'fsdsogkndfokasodnaso';

    let g = `client=${l}&mysticTime=${timestamp}&product=${d}&key=${e}`;
    let sign = crypto.createHash("md5").update(g.toString()).digest("hex");
    console.log(sign);
    return sign;
}
function getText(text) {
    oldKey = 'ydsecret://query/key/B*RGygVywfNBwpmBaZg*WT7SIOUP2T0C9WHMZN39j^DAdaZhAnxvGcCY6VYFwnHl';
    oldIv = 'ydsecret://query/iv/C@lZe2YzHtZ2CYgaXKSVfsb7Y4QWHjITPPZ0nQp87fBeJ!Iv6v^6fvi2WN@bYpJ4';

    key = Buffer.alloc(16, crypto.createHash("md5").update(oldKey).digest());
    iv = Buffer.alloc(16, crypto.createHash("md5").update(oldIv).digest());

    let decipheriv = crypto.createDecipheriv('aes-128-cbc', key, iv);

    let plaintext = decipheriv.update(text, 'base64', 'utf-8');
    plaintext += decipheriv.final('utf-8');
    return plaintext;
}

File: reverse_direction/10.有道翻译.py
import json

import requests
import time
import hashlib

# 解决execjs执行js代码报编码错误
import subprocess
from functools import partial
subprocess.Popen = partial(subprocess.Popen, encoding='utf-8')

import execjs
from fake_useragent import UserAgent

if __name__ == '__main__':
    ua = UserAgent()
    headers = {
        'User-Agent': ua.random,
        'Content-Type': 'application/x-www-form-urlencoded',
        'Referer': 'https://fanyi.youdao.com/',
        'Host': 'dict.youdao.com',
        'Origin': 'https://fanyi.youdao.com',
        'Cookie': 'OUTFOX_SEARCH_USER_ID=-867100222@10.112.57.88; OUTFOX_SEARCH_USER_ID_NCOO=894553334.5490068'
    }

    with open('10.有道翻译.js', 'r', encoding='utf-8') as f:
        js = f.read()
    ctx = execjs.compile(js)

    url = 'https://dict.youdao.com/webtranslate'
    timestamp = str(int(time.time() * 1000))
    md5 = hashlib.md5()
    md5.update(f'client=fanyideskweb&mysticTime={timestamp}&product=webfanyi&key=fsdsogkndfokasodnaso'.encode('utf-8'))
    sign = md5.hexdigest()

    data = {
        'i': 'china',
        'from': 'auto',
        'to': '',
        'domain': '0',
        'dictResult': 'true',
        'keyid': 'webfanyi',
        'sign': sign,
        'client': 'fanyideskweb',
        'product': 'webfanyi',
        'appVersion': '1.0.0',
        'vendor': 'web',
        'pointParam': 'client,mysticTime,product',
        'mysticTime': timestamp,
        'keyfrom': 'fanyi.web'
    }

    resp = requests.post(url=url, data=data, headers=headers)
    translate = ctx.call('getText', resp.text)


    json = json.loads(translate)
    print(f"翻译结果为：{json.get('translateResult')[0][0].get('tgt')}")

    resp.close()